plugins/h3/h3.cpp

```diff
@@ -0,0 +1,247 @@
+     *              initialization and termination.  Also, the system does not load
+     *              additional executable modules that are referenced by the specified module."
+     */
+
+    if (pb->size != sizeof(ECLPluginDefinitionBlock))
+        return false;
+
+    pb->magicVersion = PLUGIN_VERSION;
+    pb->version = H3_VERSION;
+    pb->moduleName = "lib_h3";
+    pb->ECL = NULL;
+    pb->flags = PLUGIN_IMPLICIT_MODULE;
+    pb->description = "ECL plugin library for uber h3\n";
+    return true;
+}
+
+#include <h3api.h>
+
+void toSetOf(bool &__isAllResult, size32_t &__lenResult, void *&__result, const H3Index *buff, int maxBuffLen)
+{
+    std::vector<H3Index> out;
+    for (int i = 0; i < maxBuffLen; ++i)
+    {
+        if (buff[i] != 0)
+            out.push_back(buff[i]);
+    }
+    __isAllResult = false;
+    __lenResult = out.size() * sizeof(H3Index);
+    __result = rtlMalloc(__lenResult);
+    memcpy(__result, &out[0], __lenResult);
+}
+
+namespace h3
+{
+
+//--------------------------------------------------------------------------------
+//                           ECL SERVICE ENTRYPOINTS
+//--------------------------------------------------------------------------------
+
+//  Indexing Functions ---
+ECL_H3_API unsigned __int64 ECL_H3_CALL index(ICodeContext *_ctx, double lat, double lon, uint32_t resolution)
+{
+    GeoCoord location;
+    location.lat = ::degsToRads(lat);
+    location.lon = ::degsToRads(lon);
+    return ::geoToH3(&location, resolution);
+}
+
+ECL_H3_API void ECL_H3_CALL center(ICodeContext *_ctx, size32_t &__lenResult, void *&__result, unsigned __int64 index)
+{
+    GeoCoord location;
+    ::h3ToGeo(index, &location);
+    __lenResult = 2 * sizeof(double);
+    __result = rtlMalloc(__lenResult);
+    double *cur = static_cast<double *>(__result);
+    *cur = ::radsToDegs(location.lat);
+    *(++cur) = ::radsToDegs(location.lon);
+}
+
+ECL_H3_API void ECL_H3_CALL boundary(ICodeContext *_ctx, size32_t &__lenResult, void *&__result, unsigned __int64 index)
+{
+    GeoBoundary boundary;
+    ::h3ToGeoBoundary(index, &boundary);
+    __lenResult = boundary.numVerts * 2 * sizeof(double);
+    __result = rtlMalloc(__lenResult);
+
+    double *cur = static_cast<double *>(__result);
+    for (int v = 0; v < boundary.numVerts; v++)
+    {
+        *cur = ::radsToDegs(boundary.verts[v].lat);
+        *(++cur) = ::radsToDegs(boundary.verts[v].lon);
+        ++cur;
+    }
+}
+
+//  Introspection ---
+ECL_H3_API uint32_t ECL_H3_CALL resolution(ICodeContext *_ctx, unsigned __int64 index)
+{
+    return ::h3GetResolution(index);
+}
+
+ECL_H3_API uint32_t ECL_H3_CALL baseCell(ICodeContext *_ctx, unsigned __int64 index)
+{
+    return ::h3GetBaseCell(index);
+}
+
+ECL_H3_API void ECL_H3_CALL toString(ICodeContext *_ctx, size32_t &lenVarStr, char *&varStr, unsigned __int64 index)
+{
+    varStr = static_cast<char *>(rtlMalloc(17));
```

Comment:
What's special about 17 ? Should it be defined as a constant somewhere?

Reply:
Its from their docs and code:  https://github.com/hpcc-systems/h3/blob/master/src/h3lib/lib/h3Index.c#L66